### DISCUSSION TITLE
release/testing/master terminology

### DISCUSSION THREAD

**Discussion by tessus** [dani-garcia/vaultwarden#4659] [General]
Moving a conversation started in an issue...

Testing is not really testing, because you only push prod ready PRs to master, which is the branch used for creating the testing containers.
So for me testing is a branch or a container to test code that is not prod ready yet. But this does not exist in this project, otherwise you would have merged several PRs already which you would like people to test.
In this project testing is a container that is the latest prod, but without a specific release tag. That's all. This is not testing in the sense of testing. ;-)

I would like to figure out whether a different approach wouldn't make more sense.

e.g. something like:

- do not label containers created from master `testing`. they don't have a release number but really are prod
- create a testing branch, from which you create `testing` containers labelled with `testing`
- push PRs to this testing branch early, so that people can test PRs that have be tested (e.g. like the camelcase PR)

**Comment by BlackDex** [comment]
I do not agree with your statement that we only push production ready PR's. Do we check for valid code, or obvious issues, yes. Does that mean that PR is totally errorless, probably not., but could be.

A lot of projects do not merge PR's which are not tested first, same here. 

Also, merging faster and breaking there `testing` tagged images, or maybe calling them `dev` in those cases and then somewhere along the line create a testing container is also weird i think.

And stuff like that is harder to maintain also.

The only workflow which could make it a bit more like you mention is if we create a `develop` branch, and only allow people to create PR's for that branch.

But then there is a need for some features to be sent to production, but not all. That is where stuff will become complicated and a hassle. If that means testing is mainly a not yet released stable, that's fine by me.

And the were enough cases where we merged PR's which in the end has some issues.

If you have a good workflow which would help here, and do not make it harder to maintain and keep differences between testing and stable without using cherry picking everytime, please share 😄

**Reply by tessus** [reply]
> A lot of projects do not merge PR's which are not tested first, same here.

I am not saying to not test at all. Additionally most projects use test frameworks where the pipeline run cargo test. The test coverage in vw is very, very low.
The camelcase PR was created with the description that most of the things work and have been tested. For covering edge cases using proper TDD would certainly help.

I rather meant that your testing containers are not for testing, but rather prod containers without a release tag. More than once you suggested people to use a testing container if they wanted X to work. Well, who installs a testing container on a prod system?

> The only workflow which could make it a bit more like you mention is if we create a develop branch, and only allow people to create PR's for that branch.

That's pretty much the same what I said. Just differently worded.

> And the were enough cases where we merged PR's which in the end has some issues.

This is true for releases as well.

> If you have a good workflow which would help here, and do not make it harder to maintain and keep differences between testing and stable without using cherry picking everytime, please share 

I could imagine a workflow like this:

- use default `develop` or `dev` branch in the repo
- PRs are opened against that branch
- testing containers (and maybe binary pre-releases are created from that branch)
- if there's an issue with a PR in dev, revert or fix
- if you decide to release (with tag or just a newer prod release without a tag), merge dev to master (type of merge TBD)

But maybe you are correct and such a workflow is not a good idea. Come to think of it, the following would probably be much better:

- add test cases to cover almost every aspect of the code (ongoing process, start with a few that can be used as a template for further tests)
- require test cases for contributions (basically enforce TDD)
- remove the `testing` label from the containers created from master

**Reply by BlackDex** [reply]
The workflow for dev you suggest is exactly what we do now but then with the main branch😄.

I'm all for better testing. That is why i was glad someone created some playwright scripts on which i hope we can build further on.

Same for more testing code for all other functions, so contributions are welcome. 

And, currently for my self, i can't quickly fix anything, because i do not have the rights in this repo right now.

The reason i say the people need to install testing is often because they want a newer web-vault.

I'm all for releasing new tagged version more often, but I'm a bit of a perfectionist my self, and that might reflect sometimes on my reasons to not approve a PR more quickly then you or others would like to see happen. 

Also take into account that if something goes wrong and a lot of people are affected, both me and dani might not  always be able to fix those items quickly. This project is still done in our spare time.
